File: schemas/publication.py
from datetime import datetime
from typing import Optional, Union, List

from pydantic import BaseModel


class PublicationSchemaBase(BaseModel):
    publication_id: int
    oai_identifier: str
    oai_identifier_num: int
    title: Optional[str]
    description: Optional[str]
    publisher: Optional[str]
    date: Optional[str]
    year: Optional[int]
    source: Optional[str]
    language: Optional[str]
    format: Optional[str]
    embedded: bool
    set_spec: Optional[str]
    is_dim_reduced: bool
    authors: Optional[List[Union["AuthorSchemaBase", "AuthorSchemaFull"]]]
    faculty_id: Optional[int]
    faculty: Optional[Union["FacultySchemaBase", "FacultySchemaFull"]]
    institute_id: Optional[int]
    institute: Optional[Union["InstituteSchemaBase", "InstituteSchemaFull"]]
    division_id: Optional[int]
    division: Optional[Union["DivisionSchemaBase", "DivisionSchemaFull"]]
    collection_id: Optional[int]

    """
    sdg_target_predictions: Optional[List[Union["SDGTargetPredictionSchemaBase", "SDGTargetPredictionSchemaFull"]]]
    clusters: Optional[List[Union["PublicationClusterSchemaBase", "PublicationClusterSchemaFull"]]]
    fact: Optional[Union["FactSchemaBase", "FactSchemaFull"]]
    summary: Optional[Union["SummarySchemaBase", "SummarySchemaFull"]]
    collection: Optional[Union["CollectionSchemaBase", "CollectionSchemaFull"]]
    """

    model_config = {
        "from_attributes": True,  # Enables ORM-style model validation
    }


class PublicationSchemaFull(PublicationSchemaBase):
    sdg_predictions: Optional[List[Union["SDGPredictionSchemaBase", "SDGPredictionSchemaFull"]]]
    dimensionality_reductions: Optional[List[Union["DimensionalityReductionSchemaBase", "DimensionalityReductionSchemaFull"]]]
    sdg_label_summary: Optional[Union["SDGLabelSummarySchemaBase", "SDGLabelSummarySchemaFull"]]

    created_at: datetime
    updated_at: datetime

    model_config = {
        "from_attributes": True, # Enables ORM-style model validation
    }

File: frontend/types/publication.ts
import type { AuthorSchemaBase, AuthorSchemaFull } from "./author";
import type { SDGLabelSummarySchemaBase, SDGLabelSummarySchemaFull } from "./sdgLabelSummary";
import type { SDGPredictionSchemaBase, SDGPredictionSchemaFull } from "./sdgPrediction";
import type { SDGTargetPredictionSchemaBase, SDGTargetPredictionSchemaFull } from "./sdgTargetPrediction";
import type { DimensionalityReductionSchemaBase } from "./dimensionalityReduction";
import type { PublicationClusterSchemaBase, PublicationClusterSchemaFull } from "./clusters/publicationCluster";
import type { FactSchemaBase, FactSchemaFull } from "./fact";
import type { SummarySchemaBase, SummarySchemaFull } from "./summary";
import type { FacultySchemaBase, FacultySchemaFull } from "./faculty";
import type { InstituteSchemaBase, InstituteSchemaFull } from "./institute";
import type { DivisionSchemaBase, DivisionSchemaFull } from "./division";
import type {CollectionSchemaBase, CollectionSchemaFull} from "./collection";

export interface PublicationSchemaBase {
  publicationId: number;
  oaiIdentifier: string;
  oaiIdentifierNum: number;
  title?: string | null;
  description?: string | null;
  publisher?: string | null;
  date?: string | null;
  year?: number | null;
  source?: string | null;
  language?: string | null;
  format?: string | null;
  embedded: boolean;
  setSpec?: string | null;
  isDimReduced: boolean;

  authors?: (AuthorSchemaBase | AuthorSchemaFull)[] | null;
  sdgLabelSummary?: (SDGLabelSummarySchemaBase | SDGLabelSummarySchemaFull)[] | null;
  sdgPredictions?: (SDGPredictionSchemaBase | SDGPredictionSchemaFull)[] | null;
  sdgTargetPredictions?: (SDGTargetPredictionSchemaBase | SDGTargetPredictionSchemaFull)[] | null;
  dimensionalityReductions?: DimensionalityReductionSchemaBase[] | null;

  clusters?: (PublicationClusterSchemaBase | PublicationClusterSchemaFull)[] | null;
  fact?: FactSchemaBase | FactSchemaFull | null;
  summary?: SummarySchemaBase | SummarySchemaFull | null;

  facultyId?: number | null;
  faculty?: (FacultySchemaBase | FacultySchemaFull)[] | null;
  instituteId?: number | null;
  institute?: (InstituteSchemaBase | InstituteSchemaFull)[] | null;
  divisionId?: number | null;
  division?: (DivisionSchemaBase | DivisionSchemaFull)[] | null;

  collectionId?: number | null;
  collection?: CollectionSchemaBase | CollectionSchemaFull | null;
}

export interface PublicationSchemaFull extends PublicationSchemaBase {
  createdAt: string;
  updatedAt: string;
}
